File: py_ss/pendulum_with_torque.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import matplotlib.pyplot as plt
import numpy as np

import blocks, helper
from blocks import N as N

class SSModel(blocks.Submodel):
    def __init__(self):
        super().__init__('pendulum')

        phi = N('phi')
        dphi = N('dphi')
        tau = N('tau')
        m = N('m')
        g = N('g')
        l = N('l')

        with self:
            blocks.Integrator('dphi', iport='d2phi', oport=dphi, x0=np.pi/4)
            blocks.Integrator('phi', iport=dphi, oport=phi)
            blocks.Function('sin(phi)', act_func=lambda t, x: np.sin(x),
                            iport=phi)
            blocks.MulDiv('g/l', operations='**/', iports=['-', g, l],
                          oport=-1)
            blocks.MulDiv('', '*///', iports=[tau, m, l, l], oport=-2)
            blocks.AddSub('', '+-', [-2, -1], 'd2phi')

def main():
    parameters = {
        'm': 0.2,
        'l': 0.1,
        'g': 9.81,
        'tau': 0.13,
        }

    model = SSModel()
    history = helper.run(model, T=np.arange(0.0, 5.0, 0.01),
                         parameters=parameters)

    plt.figure()    
    plt.subplot(2, 1, 1); plt.plot(history['t'], history['phi'])
    plt.ylabel('phi')
    plt.subplot(2, 1, 2); plt.plot(history['t'], history['dphi'])
    plt.xlabel('t'); plt.ylabel('dphi')
    plt.show()

if __name__ == '__main__':
    main()

File: cpp_ss/sandbox/pendulum_with_torque.cpp

#include <iostream>
#include <math.h>
#include <vector>

#include "blocks.hpp"
#include "helper.hpp"
#include "solver.hpp"
#include "gp-ios.hpp"

using namespace blocks;

class SSModel : public Submodel
{
public:
    SSModel() : Submodel("pendulum")
    {
        Node   phi(  "phi");
        Node  dphi( "dphi");
        Node d2phi("d2phi");
        Node tau("tau");
        Node m("m");
        Node g("g");
        Node l("l");

        enter();
        {
            new Integrator("dphi", d2phi, dphi, M_PI_4);
            new Integrator("phi", dphi, phi);
            new Function("sin(phi)",
                [](double /*t*/, const Value& x) -> Value
                {
                    return x.sin();
                }, phi);
            // new Sin("sin(phi)", phi);
            new MulDiv("g/l", "**/", {"", g, l}, -1);
            new MulDiv("", "*///", {tau, m, l, l}, -2);
            new AddSub("", "+-", {-2, -1}, "d2phi");
        }
        exit();
    }
};

int main()
{
    NodeValues parameters = {
        {  "m", 0.2 },
        {  "l", 0.1 },
        {  "g", 9.81},
        {"tau", 0.13},
        };

    auto model = SSModel();
    auto history = run(model,
        [](uint k, double& t) -> bool
        {
            return arange(k, t, 0, 5, 0.01);
        },
        nullptr, parameters, rk4);

    Gnuplot gp;
	gp << "set xrange [0:500]\n";
    gp << "set yrange [-8:8]\n";
	gp << "plot" << gp.file1d(history["phi"]) << "with lines title 'phi',"
	    << gp.file1d(history["dphi"]) << "with lines title 'dphi'\n";

    return 0;
}
